File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.py
class Solution:
    def findOriginalArray(self, changed: List[int]) -> List[int]:
        if(len(changed) % 2 == 1):
            return []
        maxN = 100001
        count = [0] * maxN
        
        maxi = -1
        for num in changed:
            count[num] += 1
            maxi = max(maxi,num)
            
            
        res = []
        for num in range(maxi,0,-1):
            mini = min(count[num],count[num//2])
            if(count[num] == 0):
                continue
            if(num % 2 != 0 and count[num] != 0):
                return []
            temp = [num//2]*mini
            count[num] -= mini
            count[num//2] -= mini
            res.extend(temp)
            if(count[num] != 0):
                return []
        
        if(count[0] % 2 != 0):
            return []
        
        temp = [0] * (count[0]//2)
        res.extend(temp)
            
        
        return res
            
                
            
        

File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.cpp
class Solution {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        if((int)changed.size() % 2 != 0) {
            return {};
        }
        vector<int> c(100001,0);
        int maxi = -1;
        for (auto x : changed) {
            c[x]++;
            maxi = max(maxi,x);
        }
        vector<int> ans;
        for (int i = maxi; i > 0; i--) {
            if(c[i] == 0)
                continue;
            if(i%2 != 0)
                return {};
            int mini = min(c[i],c[i/2]);
            for(int j = 0; j < mini; j++) {
                ans.push_back(i/2);
            }
            c[i] -= mini;
            c[i/2] -= mini;
            if(c[i] != 0) {
                return {};
            }
        }
        if(c[0] % 2 != 0) {
            return {};
        }
        for(int i = 0; i < c[0]/2; i++) {
            ans.push_back(0);
        }
        return ans;
    }
};